### PR TITLE
Add IBM MQ Mixin

### PR DESCRIPTION
**Pull request by armstrmi**
Adds dashboards and alerts for IBM MQ.

Dashboards:

 IBM MQ cluster overview
 IBM MQ queue overview
 IBM MQ queue manager overview
 IBM MQ topics overview

Alerts:

IBMMQStaleMessages
IBMMQStaleMessages
IBMMQLowDiskSpace
IBMMQHighQueueManagerCpuUsage

## IBM MQ queue manager overview

image
image

## IBM MQ queue overview

image

## IBM MQ cluster overview

image

## IBM MQ topics overview

image

### CODE REVIEW

File: ibm-mq-mixin/config.libsonnet [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    dashboardPeriod: 'now-1h',
+    dashboardTimezone: 'default',
+    dashboardRefresh: '1m',
+
+    //alerts thresholds
+    alertsExpiredMessages: 2,  //count
+    alertsStaleMessagesSeconds: 300,  //seconds
+    alertsLowDiskSpace: 5,  //percentage: 0-100
+    alertsHighQueueManagerCpuUsage: 85,  //percentage: 0-100
+    enableLokiLogs: true,
```
Comment (JonathanWamsley) [inline ibm-mq-mixin/config.libsonnet:13]:
```suggestion

    enableLokiLogs: true,
```

Nit: New line to separate from alerts

File: ibm-mq-mixin/.lint [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    - panel: "Topic subscribers"
+    - panel: "Topic publishers"
+    - panel: "Subscription messages received"
+    - panel: "Active listeners"
+    - panel: "Active connections"
+    - panel: "Published messages"
+    - panel: "Commits"
+    - panel: "Expired messages"
+    - panel: "Queue operations"
+    - panel: "Operations"
```
Comment (JonathanWamsley) [inline ibm-mq-mixin/.lint:33]: I think we normally have operation metrics with a unit of ops or operations.

File: ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet [hunk 3]
```diff
@@ -0,0 +1,1287 @@
+local g = (import 'grafana-builder/grafana.libsonnet');
```
Comment (JonathanWamsley) [inline ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet:1]: I believe this import is not used here or in the other .libsonnet dashboards and can be removed.

File: ibm-mq-mixin/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+- IBMMQStaleMessages
+- IBMMQStaleMessages
+- IBMMQLowDiskSpace
+- IBMMQHighQueueManagerCpuUsage
+
+## IBM MQ cluster overview 
+
+The IBM MQ cluster overview dashboard provides details on the number and status of clusters, queue managers, queues, and topics within and IBM MQ cluster. This also includes information about the amount of time it takes for a messsage to get through a transmission queue.
+
+# screenshots
```
Comment (schmikei) [inline ibm-mq-mixin/README.md:23]:
```suggestion
# Screenshots
```

File: ibm-mq-mixin/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+![]()
```
Comment (schmikei) [inline ibm-mq-mixin/README.md:24]: Speaking of screenshots, I think the ones in the PR need more load in order to come across better. I.e. more operations and an actual `mq_cluster` and not the `<mq_cluster>`

File: ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet [hunk 6]
```diff
@@ -0,0 +1,1294 @@
...
+    reduceOptions: {
+      calcs: [
+        'lastNotNull',
+      ],
+      fields: '',
+      values: false,
+    },
+    textMode: 'auto',
+  },
+  pluginVersion: '10.0.1-cloud.3.f250259e',
```
Comment (schmikei) [inline ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet:66]: Do we normally include the pluginVersion?

**Review by schmikei — CHANGES_REQUESTED** [review]
Just a few questions :)

File: ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet (hunk 6 above)
Comment (armstrmi) [reply ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet:66]: I am unsure honestly, I see it in a few other mixin's, 
image

File: ibm-mq-mixin/README.md (hunk 5 above)
Comment (armstrmi) [reply ibm-mq-mixin/README.md:24]: I can for sure do more operations, but I will say in that screenshot, we are using an MQ cluster named `INVENTORY` . Given the screenshot is a little rough. We were aiming to retake some shots once we got some more eyes on the PR and if there was any more feedback.

**Review by schmikei — APPROVED** [review]

File: ibm-mq-mixin/README.md (hunk 5 above)
Comment (Dasomeone) [reply ibm-mq-mixin/README.md:24]: Would definitely appreciate a bit more load, shows up much nicer in our docs that way too :)

**Review by Dasomeone — CHANGES_REQUESTED** [review]
Apologies on the delay for this review, both forgot and wanted to make sure we'd caught up internally on style consistency first.

Left a bunch of style-oriented comments on the individual dashboards, but all the base queries, units, etc all seem great! 

Given the style feedback I'm holding off on uploading the screenshots just yet.

File: ibm-mq-mixin/dashboards/cluster-overview.libsonnet
[the remark this answers is not in the document]
Comment (gaantunes) [reply ibm-mq-mixin/dashboards/cluster-overview.libsonnet:1]: I would love to see some color on the `status` panels, to easily show if it is running (green), down (red), stale (yellow), this kind of categorization.

File: ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet [hunk 7]
```diff
@@ -0,0 +1,1294 @@
...
+      showLegend: true,
+    },
+    tooltip: {
+      mode: 'single',
+      sort: 'none',
+    },
+  },
+};
+
+local queueManagerStatusPanel = {
```
[the remark this answers is not in the document]
Comment (gaantunes) [reply ibm-mq-mixin/dashboards/queue-manager-overview.libsonnet:253]: Agreed here with Emily, would just put the stats panels on the left, feels better to look at.

**Comment by gaantunes** [comment]
I am having a hard time undertanding a few of the legends of some panels:

Cluster overview - transmission queue time - what long and short means? Also, the data shown on the graph misses the prefix for some reason.
QM Overview - CPU Usage, Disk Usage, Commits, Publish Throughput, Plublished Messages, Expired Messages, Log Latency, Log Usage - Those have the same legend for both data being shown, so it is hard to figure out what I am seeing stacked.

**Comment by armstrmi** [comment]
> I am having a hard time understanding the legends of some panels:
> 
> Cluster overview - transmission queue time - what long and short means? Also, the data shown on the graph misses the prefix for some reason. QM Overview - CPU Usage, Disk Usage, Commits, Publish Throughput, Plublished Messages, Expired Messages, Log Latency, Log Usage - Those have the same legend for both data being shown, so it is hard to figure out what I am seeing stacked.

@gaantunes  Yea for sure, so the difference between long and short metric. 

The long metric typically measures the total time taken for messages to be transmitted over the channel, from the moment they are put on the XMITQ until they are successfully sent and acknowledged by the remote end of the channel.

The short metric represents a shorter or more recent timeframe for the transmission time. It may be an average, minimum, or maximum time taken to transmit messages over the channel in recent intervals or for a specific number of messages. 

The reason the prefix was missing in those particular screenshots is that we were using a metrics generator for the sake of load on those metrics, unfortunately forgot to add the `channel` label on those metrics, therefore the legend `{{channel}} - short` `... - long` didn't show those labels.

Similarly, for those other panels in the QM overview dashboard, that is an instance where we were sending metrics from the metric generator while also getting metrics from the actual IBM MQ instance, therefore that's why you see alot of the duplicated legend on the bottom. Nothing is really being stacked, its two separate sets of data being sent to the grafana instance.

File: ibm-mq-mixin/dashboards/cluster-overview.libsonnet
[the remark this answers is not in the document]
Comment (Dasomeone) [reply ibm-mq-mixin/dashboards/cluster-overview.libsonnet:1]: Yes! Threshold values would be great for the status panels!

File: ibm-mq-mixin/dashboards/topics-overview.libsonnet [hunk 8]
```diff
@@ -0,0 +1,634 @@
...
+          ),
+          template.new(
+            'mq_cluster',
+            promDatasource,
+            'label_values(ibmmq_topic_messages_received{job=~"$job"},mq_cluster)',
+            label='MQ cluster',
+            refresh=2,
+            includeAll=false,
+            multi=false,
+            allValues='',
```
Comment (Dasomeone) [inline ibm-mq-mixin/dashboards/topics-overview.libsonnet:583]: Noticed an issue on this dashboard, the default value for the "MQ Cluster" selector wasn't loading properly, and so no data was shown.

The value here should probably be as follows to allow selecting properly?
```suggestion
            includeAll=true,
            multi=true,
            allValues='.+',
```

File: ibm-mq-mixin/dashboards/cluster-overview.libsonnet [hunk 9]
```diff
@@ -695,14 +704,14 @@ local transmissionQueueTimePanel = {
...
-          queueOperationsPanel { gridPos: { h: 16, w: 12, x: 12, y: 0 } },
-          topicsPanel { gridPos: { h: 8, w: 6, x: 0, y: 8 } },
-          queuesPanel { gridPos: { h: 8, w: 6, x: 6, y: 8 } },
-          clusterStatusPanel { gridPos: { h: 5, w: 24, x: 0, y: 16 } },
-          queueManagerStatusPanel { gridPos: { h: 4, w: 24, x: 0, y: 21 } },
-          transmissionQueueTimePanel { gridPos: { h: 8, w: 24, x: 0, y: 25 } },
+          clustersPanel { gridPos: { h: 7, w: 4, x: 0, y: 0 } },
+          queueManagersPanel { gridPos: { h: 7, w: 4, x: 4, y: 0 } },
+          topicsPanel { gridPos: { h: 7, w: 4, x: 8, y: 0 } },
+          queuesPanel { gridPos: { h: 7, w: 4, x: 12, y: 0 } },
```
[the remark this answers is not in the document]
Comment (armstrmi) [reply ibm-mq-mixin/dashboards/cluster-overview.libsonnet:710]: Yea, I tried making them 4x4 and it seems like they turn out to be more rectangles
image

[the remark this answers is not in the document]
Comment (Dasomeone) [reply ibm-mq-mixin/dashboards/cluster-overview.libsonnet:710]: Yeah, it's tied to your resolution and aspect ratio, so when we say 4x4 it's just the grid units. They'll show as rectangular on a wide-angle monitor, etc

**Comment by gaantunes** [comment]
> The long metric typically measures the total time taken for messages to be transmitted over the channel, from the moment they are put on the XMITQ until they are successfully sent and acknowledged by the remote end of the channel.
> 
> The short metric represents a shorter or more recent timeframe for the transmission time. It may be an average, minimum, or maximum time taken to transmit messages over the channel in recent intervals or for a specific number of messages.

Got it! Thanks for explaining. In that case, if we can put this as information on this panel this would be great!.

**Comment by armstrmi** [comment]
> > The long metric typically measures the total time taken for messages to be transmitted over the channel, from the moment they are put on the XMITQ until they are successfully sent and acknowledged by the remote end of the channel.
> > The short metric represents a shorter or more recent timeframe for the transmission time. It may be an average, minimum, or maximum time taken to transmit messages over the channel in recent intervals or for a specific number of messages.
> 
> Got it! Thanks for explaining. In that case, if we can put this as information on this panel this would be great!.

Would that be an update in the description?

**Comment by gaantunes** [comment]
> > > The long metric typically measures the total time taken for messages to be transmitted over the channel, from the moment they are put on the XMITQ until they are successfully sent and acknowledged by the remote end of the channel.
> > > The short metric represents a shorter or more recent timeframe for the transmission time. It may be an average, minimum, or maximum time taken to transmit messages over the channel in recent intervals or for a specific number of messages.
> > 
> > 
> > Got it! Thanks for explaining. In that case, if we can put this as information on this panel this would be great!.
> 
> Would that be an update in the description?

Yes, just add a short description of what long and short means.

**Comment by armstrmi** [comment]
Updated @gaantunes

**Comment by gaantunes** [comment]
> Updated @gaantunes

Thank you!

**Review by gaantunes — APPROVED** [review]
LGTM, great work!

**Review by Dasomeone — APPROVED** [review]
This looks fantastic, thanks for all the work on this! 😄

**Pull request merged**